volley/volley.py

```diff
@@ -0,0 +1,168 @@
+    lng_a = self.normalize_radians(lng_a)
+    lng_b = self.normalize_radians(lng_b)
+    
+    # Here, normalize to [-pi, pi) for longitude
+    return self.normalize_radians((lng_a + lng_b) / 2, -1 * math.pi, math.pi)
+
+
+
+  # Helper for weighted_spherical_mean, defined in Volley paper
+  #
+  # params:
+  #   weight: weight for interpolation
+  #   loc_a: lat/lng for location A
+  #   loc_b: lat/lng for location B
+  def interp(self, weight, loc_a, loc_b):
+
+    lat_a = self.convert_lat_to_radians(loc_a[0])
+    lng_a = self.convert_lng_to_radians(loc_a[1])
+    lat_b = self.convert_lat_to_radians(loc_b[0])
+    lng_b = self.convert_lng_to_radians(loc_b[1])
+
+    d_first = math.cos(lat_a) * math.cos(lat_b)
+    d_second = math.sin(lat_a) * math.sin(lat_b) * math.cos(lng_b - lng_a)
+    d = math.acos(d_first + d_second)
+
+    gamma_numerator = math.sin(lat_b) * math.sin(lat_a) * math.sin(lng_b - lng_a)
+    gamma_denominator = math.cos(lat_a) - (math.cos(d) * math.cos(lat_b))
+    gamma = math.atan2(gamma_numerator, gamma_denominator)
+
+    beta_numerator = math.sin(lat_b) * math.sin(weight * d) * math.sin(gamma)
+    beta_denominator = math.cos(weight * d) - (math.cos(lat_a) * math.cos(lat_b))
+    beta = math.atan2(beta_numerator, beta_denominator)
+
+    lat_c_first = math.cos(weight * d) * math.cos(lat_b)
+    lat_c_second = math.sin(weight * d) * math.sin(lat_b) * math.cos(gamma)
+    lat_c = math.acos(lat_c_first + lat_c_second)
+    lat_c = self.convert_lat_to_degrees(lat_c)
+
+    # Find an average of coming from either direction for antipodal nodes
+    lng_c_1 = lng_b - beta
+    lng_c_2 = lng_a + beta
+    lng_c = self.find_avg_lng(lng_c_1, lng_c_2)
+    lng_c = self.convert_lng_to_degrees(lng_c)
+
+    return (lat_c, lng_c)
+
+  # Recursive helper for weighted_spherical_mean
+  #
+  # params:
+  #   weights: a list of weights
+  #   locations: a list of latitude/longitude tuples for clients, has same cardinality as weights
+  def weighted_spherical_mean_helper(self, total_weight, weights, locations):
+    if len(weights) != len(locations):
+      raise ValueError('Weights and locations must have the same length.')
+    
+    length = len(weights)
+
+    print '------ Weighted spherical mean -------- '
+    print 'Length: ' + str(length)
+    print 'WEIGHTS: ' + str(weights)
+    print 'LOCATIONS: ' + str(locations)
+
+    current_weight = float(weights.pop())
+    weight = current_weight / total_weight
+    location = locations.pop()
+
+    print 'current_weight: ' + str(weight)
+
+    if length == 1:
+      return location
+
+    return self.interp(weight, location, self.weighted_spherical_mean_helper(total_weight, weights, locations))
+
+
+  # Find the weighted spherical mean location for a data item
+  #
+  # params:
+  #   uuid: uuid of the data item
+  def weighted_spherical_mean(self, uuid):
+    # Find results for each source entity
+
+    # TODO: WHAT KIND OF REQUESTS? GET ONLY SUCCESSFUL READ REQUESTS?
```

Comment:
Still a TODO? or it is already taken care of.

I would say that we should count all `read` requests, even it fails.